src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs

```diff
@@ -0,0 +1,399 @@
+            {
+                _observer = observer;
+                _source = source;
+                _timeSelector = timeSelector;
+
+                _scheduler = scheduler ?? GlobalConfig.DefaultScheduler;
+
+                _expirationDueTimesByKey = new();
+                _proposedExpirationsQueue = new();
+                _removedItemsBuffer = new();
+
+                _sourceSubscription = source
+                    .Connect()
+                    // It's important to set this flag outside the context of a lock, because it'll be read outside of lock as well.
+                    .Finally(() => _hasSourceCompleted = true)
+                    .Synchronize(SynchronizationGate)
+                    .SubscribeSafe(Observer.Create<IChangeSet<TObject, TKey>>(
+                        onNext: OnSourceNext,
+                        onError: OnSourceError,
+                        onCompleted: OnSourceCompleted));
+            }
+
+            public void Dispose()
+            {
+                lock (SynchronizationGate)
+                {
+                    _sourceSubscription.Dispose();
+
+                    TryCancelNextScheduledManagement();
+                }
+            }
+
+            protected IScheduler Scheduler
+                => _scheduler;
+
+            // Instead of using a dedicated _synchronizationGate object, we can save an allocation by using any object that is never exposed to public consumers.
+            protected object SynchronizationGate
+                => _expirationDueTimesByKey;
+
+            protected abstract DateTimeOffset? GetNextManagementDueTime();
+
+            protected DateTimeOffset? GetNextProposedExpirationDueTime()
+                => _proposedExpirationsQueue.Count is 0
+                    ? null
+                    : _proposedExpirationsQueue[0].DueTime;
+
+            protected abstract void OnExpirationsManaged(DateTimeOffset dueTime);
+
+            private void ClearExpiration(TKey key)
+                // This is what puts the "proposed" in _proposedExpirationsQueue.
+                // Finding the position of the item to remove from the queue would be O(log n), at best,
+                // so just leave it and flush it later during normal processing of the queue.
+                => _expirationDueTimesByKey.Remove(key);
+
+            private void ManageExpirations()
+            {
+                // This check is needed, to make sure we don't try and call .Edit() on a disposed _source,
+                // since the scheduler only promises "best effort" to cancel a scheduled action.
+                // It's safe to skip locking here becuase once this flag is set, it's never unset.
+                if (_hasSourceCompleted)
+                    return;
+
+                // Putting the entire management process here inside a .Edit() call for a couple of reasons:
+                //  - It keeps the edit delegate from becoming a closure
```

Comment:
It's still a closure though.  It captures all the members variables... Which might compile to just capturing `this`, but it captures something.

We should look at putting a `TState<>` parameter to the edit method like you suggested.

Reply:
It's not QUITE the same as a closure, because like you said, the only state that gets captured is `this`. Unless I'm much mistaken, what this compiles to is just moving the delegate to an instance method on the class, and then the delegate becomes an allocation per-instance of the class, rather than per call to `.Edit(). IIRC, this was discussed in a blog post discussing general delegate enhancements in .NET 6 and .NET 7. Yes, parameterizing the state to make it a `static` delegate would be even better.

Reply:
This is the article I was referring to, and after re-reading, and doing a little ILSpy, I confirmed you're correct. There's no closure, but there is still an allocation, per call to `.Edit()`. Fixing now...